csharp/src/Ice/BufferedTransceiver.cs

```diff
@@ -0,0 +1,178 @@
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ZeroC.Ice
+{
+    /// <summary>The BufferedReadTransceiver is a wrapper around the ITransceiver interface to provide buffered
+    /// receive. This helps to limit the number of operating system Receive calls when the user needs to read
+    /// only few bytes before reading more (typically to read a frame header) by receiving the data in a small
+    /// buffer. It's similar to the C# System.IO.BufferedStream class. It's used to implement Slic and WebSocket.
+    /// </summary>
+    internal class BufferedReadTransceiver : ITransceiver
+    {
+        public Socket? Socket => Underlying.Socket;
+        internal ITransceiver Underlying { get; }
+
+        // The buffered data.
+        private ArraySegment<byte> _buffer;
+
+        public void CheckSendSize(int size) => Underlying.CheckSendSize(size);
+
+        public ValueTask CloseAsync(Exception exception, CancellationToken cancel) =>
+            Underlying.CloseAsync(exception, cancel);
+
+        public ValueTask DisposeAsync() => Underlying.DisposeAsync();
+
+        public ValueTask InitializeAsync(CancellationToken cancel) => Underlying.InitializeAsync(cancel);
+
+        public ValueTask<ArraySegment<byte>> ReceiveAsync(CancellationToken cancel) =>
+            throw new InvalidOperationException();
+
+        public async ValueTask<int> ReceiveAsync(ArraySegment<byte> buffer, CancellationToken cancel = default)
+        {
+            int received = 0;
+            if (_buffer.Count > 0)
+            {
+                // If there's still data buffered for the payload, consume the buffered data.
+                int length = Math.Min(_buffer.Count, buffer.Count);
+                ReadOnlyMemory<byte> buffered = await ReceiveAsync(length, cancel).ConfigureAwait(false);
+                buffered.CopyTo(buffer);
+                received = length;
+            }
+
+            // Then, read the reminder from the underlying transport.
+            if (received < buffer.Count)
+            {
+                received += await Underlying.ReceiveAsync(buffer.Slice(received), cancel).ConfigureAwait(false);
+            }
+            return received;
+        }
+
+        /// <summary>Returns buffered data. If there's no buffered data, the buffer is filled using the underlying
+        /// transceiver to receive additional data. The method returns when the buffer contains at least byteCount
+        /// data. If byteCount is 0, it returns all the buffered data.</summary>
+        /// <param name="byteCount">The number of bytes of buffered data to return. It can be set to null to get all
+        /// the buffered data.</param>
+        /// <param name="cancel">A cancellation token that receives the cancellation requests.</param>
+        /// <return>The buffered data.</return>
+        public async ValueTask<ReadOnlyMemory<byte>> ReceiveAsync(int byteCount, CancellationToken cancel = default)
```

Comment:
I'm not too keen on changing this... It's an internal API here and it has the advantage that the caller can directly decode the data from the buffered data rather than having to use an intermediary working buffer which would require an additional copy. If you feel strongly about it, I'll fix it but I'd prefer to not do this in this PR as this requires to review all the usage of this method, in particular for the WS transceiver.